src/overlay/Peer.cpp

```diff
@@ -1840,6 +1951,33 @@ Peer::recvSurveyResponseMessage(StellarMessage const& msg)
         msg, shared_from_this());
 }
 
+void
+Peer::queueTxHashAndMaybeTrim(Hash const& hash)
+{
+    mIncomingTxHashes.push_back(hash);
+    while (mIncomingTxHashes.size() > mApp.getHerder().getMaxQueueSizeOps())
+    {
+        mIncomingTxHashes.pop_front();
+    }
+}
+
+void
+Peer::recvFloodAdvert(StellarMessage const& msg)
+{
+    for (Hash const& hash : msg.floodAdvert().txHashes)
+    {
+        mPeerMetrics.mTxHashReceived++;
+        queueTxHashAndMaybeTrim(hash);
+    }
+    mApp.getOverlayManager().demand();
```

Comment:
Oh, I missed this in the last iteration, but I don't think this works. Every time an advert is queued, demand functionality is triggered right away, which may result in a lot of small demands sent out. This probably creates instabilities as we can't properly accumulate hashes and round robin among peers. Instead, in the `demand` function, you should schedule the timer regardless of whether you have hashes or not. If the timer fires and there are no hashes, you just no-op and schedule the timer again.

Reply:
(then you don't even need mWaitingToDemand)

Reply:
I just looked at the Tracy capture, and indeed it looks like most demands are triggered by receiving an advert, not by the timer

Reply:
Updated.